[jit] support grad and data attribute for tensor in script

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23842 [jit] support grad and data attribute for tensor in script**
* #23841 [jit] add initial support for sparse tensors
* #23639 [jit] support dict augment assign in script

Summary:
Add the grad and data attribute for tensor type in TorchScript, as part
of effort of support jittable training loop, we will frequently update
the .grad and sometimes access the .data attribute

not quite sure we should just disallow .data and encourage user to use
detach() instead

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23842

Differential Revision: [D16683556](https://our.internmc.facebook.com/intern/diff/D16683556)

wanchaol committed:
[jit] support grad and data attribute for tensor in script

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] support grad and data attribute for tensor in script"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] support grad and data attribute for tensor in script"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D16683556](https://our.internmc.facebook.com/intern/diff/D16683556)

zdevito commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -567,6 +567,19 @@ RegisterOperators reg(
+         },
+         aliasAnalysisFromSchema()),
+     Operator(
+         "prim::data(Tensor a) -> Tensor",
```
Alias information is wrong here. `(Tensor(b) a) -> Tensor(b)`

zdevito commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -567,6 +567,19 @@ RegisterOperators reg(
          },
          aliasAnalysisFromSchema()),
+     Operator(
+         "prim::grad(Tensor a) -> Tensor",
```
Alias information is wrong here `(Tensor a) -> Tensor(*)`. It is similar to looking up a tensor from a list.

zdevito reviewed (approved):
Please add a PR summary with context. Leaving out context makes it really hard to debug issues in the future.

wanchaol committed:
Update on "[jit] support grad and data attribute for tensor in script"


Summary:
Add the grad and data attribute for tensor type in TorchScript, as part
of effort of support jittable training loop, we will frequently update
the .grad and sometimes access the .data attribute

not quite sure we should just disallow .data and encourage user to use
detach() instead

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23842

Differential Revision: [D16683556](https://our.internmc.facebook.com/intern/diff/D16683556)

eellison commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -567,6 +567,19 @@ RegisterOperators reg(
+         aliasAnalysisFromSchema()),
+     Operator(
+         "prim::data(Tensor(b) a) -> Tensor(b)",
+         noop,
```
Why is this a noop? 
```
> a = torch.tensor(.5)
> a is a.data
False
```

wanchaol replied:
this is because python binds are different from the underlying C++, https://github.com/pytorch/pytorch/blob/c5482e33e9ef59af6c18059378964a00036d95aa/torch/csrc/autograd/python_variable.cpp#L235 python side re-wrap the underlying `at::Tensor` again in a new THPVariable, so the identity check in python is different, but the underlying C++ `at::Tensor` (also in our interpreter) is still the same.

wanchaol committed:
rebase and address comments on "[jit] support grad and data attribute for tensor in script"


Summary:
Add the grad and data attribute for tensor type in TorchScript, as part
of effort of support jittable training loop, we will frequently update
the .grad and sometimes access the .data attribute

not quite sure we should just disallow .data and encourage user to use
detach() instead

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23842

Differential Revision: [D16683556](https://our.internmc.facebook.com/intern/diff/D16683556)
